src/plots/geo/geo2.js

```diff
@@ -0,0 +1,612 @@
+                layer.append('g').classed('choroplethlayer', true);
+            } else if(d === 'frontplot') {
+                layer.append('g').classed('scatterlayer', true);
+            } else if(isTopoLayer(d)) {
+                basePaths[d] = layer.append('path');
+            }
+        });
+
+    join.order();
+
+    join.each(function(d) {
+        var path = basePaths[d];
+        var adj = constants.layerNameToAdjective[d];
+
+        if(d === 'frame') {
+            path.datum(constants.sphereSVG);
+        } else if(isTopoLayer(d)) {
+            path.datum(topojsonFeature(topojson, topojson.objects[d]));
+        } else if(isAxisLayer(d)) {
+            path.datum(makeGraticule(d, geoLayout))
+                .attr('fill', 'none')
+                .call(Color.stroke, geoLayout[d].gridcolor)
+                .call(Drawing.dashLine, '', geoLayout[d].gridwidth);
+        }
+
+        if(constants.fillLayers.indexOf(d) !== -1) {
+            path.attr('stroke', 'none')
+                .call(Color.fill, geoLayout[adj + 'color']);
+        } else if(constants.lineLayers.indexOf(d) !== -1) {
+            path.attr('fill', 'none')
+                .call(Color.stroke, geoLayout[adj + 'color'])
+                .call(Drawing.dashLine, '', geoLayout[adj + 'width']);
+        }
+    });
+};
+
+proto.updateDims = function(fullLayout, geoLayout) {
+    var b = this.bounds;
+    var hFrameWidth = (geoLayout.framewidth || 0) / 2;
+
+    var l = b[0][0] - hFrameWidth;
+    var t = b[0][1] - hFrameWidth;
+    var w = b[1][0] - l + hFrameWidth;
+    var h = b[1][1] - t + hFrameWidth;
+
+    Drawing.setRect(this.clipRect, l, t, w, h);
+
+    this.bgRect
+        .call(Drawing.setRect, l, t, w, h)
+        .call(Color.fill, geoLayout.bgcolor);
+
+    this.xaxis._offset = l;
+    this.xaxis._length = w;
+
+    this.yaxis._offset = t;
+    this.yaxis._length = h;
+};
+
+proto.updateFx = function(fullLayout, geoLayout) {
+    var _this = this;
+    var framework = _this.framework;
+    var gd = _this.graphDiv;
+    var dragMode = fullLayout.dragmode;
+
+    if(_this.isStatic) return;
+
+    function zoomReset() {
+        var viewInitial = _this.viewInitial;
+        var updateObj = {};
+
+        for(var k in viewInitial) {
+            updateObj[_this.id + '.' + k] = viewInitial[k];
+        }
+
+        Plotly.relayout(gd, updateObj);
+        gd.emit('plotly_doubleclick', null);
+    }
+
+    function invert(lonlat) {
+        return _this.projection.invert([
+            lonlat[0] + _this.xaxis._offset,
+            lonlat[1] + _this.yaxis._offset
+        ]);
+    }
+
+    if(dragMode === 'pan') {
+        _this.bgRect.node().onmousedown = null;
+        framework.call(createGeoZoom(_this, geoLayout));
+        framework.on('dblclick.zoom', zoomReset);
+    }
+
+    framework.on('mousemove', function() {
+        var lonlat = _this.projection.invert(d3.mouse(this));
+
+        if(!lonlat || isNaN(lonlat[0]) || isNaN(lonlat[1])) return;
```

Comment:
In `geo_orthographic` if I mouse off the globe while hovered on a point, this block gets hit, and I don't get an unhover event. I thought I'd be able to fix it by adding `Fx.loneUnhover` before the `return` here but that didn't seem to work. Can you take a look @etpinard ?
![screen shot 2017-09-22 at 12 06 03 pm](https://user-images.githubusercontent.com/2678795/30753904-c08e0f30-9f8e-11e7-96b2-596ebc927d72.png)

Reply:
And very often the hover tip still shows after I mouse out this way. Every once in a while I actually get a hover tip for a totally different element that I haven't ever hovered on...

Reply:
Can you check if that's also true in `latest`?

Reply:
And while we're on the topic, I notice that hover tips are still present while dragging the projection - ideally they should disappear.

Reply:
Behavior on master is similar, yes, though the second issue (hover tip for a totally different element) seems to be new:
![kapture 2017-09-22 at 12 29 42](https://user-images.githubusercontent.com/2678795/30754723-d83523fa-9f91-11e7-96d3-c69e14747e27.gif)

Reply:
> And very often the hover tip still shows after I mouse out this way.

This one was easy, using `dragElement.unhover` instead of `Fx.loneUnhover` did the trick in 9bb7d91 Unfortunately, I can get the test I wrote working on CI.

>  (hover tip for a totally different element) seems to be new

This turned out to be caused by two separate bugs:

-https://github.com/plotly/plotly.js/pull/2030/commits/1c26ec4d12f542f289634ecf6e8bae049d4e73f5 made sure that `lon`/`lat` coordinate are converted correctly to px in scattergeo/hover.js
- https://github.com/plotly/plotly.js/pull/2030/commits/651399e430dfdff6f91b3c06cebeba07898d82b7 is a little more subtle - we must make sure that location polygons that cross the antimeridian have their negative longitude offset by 360 degrees so that `polygon.contains` doesn't get confuse. I found that only Russia and Fiji had such polygons.

So now, hovering over the North Pole on `geo_orthographic` gives:

![peek 2017-09-27 10-34](https://user-images.githubusercontent.com/6675409/30919879-a494c2ca-a370-11e7-98fe-f6924b0cde72.gif)

Reply:
That was quite a deep dive - nicely done, and thanks for tacking on the first bug even though it wasn't new.

Reply:
With https://github.com/plotly/plotly.js/pull/2030/commits/190c7338b3f2a6271d54bf90129129374e5596f8 choropleth hover now works over Antarctica. Is wasn't easy, but I think I found a solid way to do this

![peek 2017-09-27 18-32](https://user-images.githubusercontent.com/6675409/30941196-3c209c72-a3b2-11e7-8f36-c203413eabdc.gif)

Reply:
Wow OK - Antarctica is especially annoying because, at least the way you did this (and I don't see a better way, this looks great) we *can't* envision fixing this at the topojson level, or we'd draw a spurious ray to the pole.

Reply:
> we can't envision fixing this at the topojson level, or we'd draw a spurious ray to the pole.

The special patches added in `choropleth/plot.js` can't _just_ be moved to the topojson creations step. Look what happens when longitudes [-180, 0] are mapped to [180, 360]:

![peek 2017-09-28 09-40](https://user-images.githubusercontent.com/6675409/30969832-893989fe-a431-11e7-9ba1-df3f70096bf9.gif)

which I think is due to d3 [optimising](https://bl.ocks.org/mbostock/5735770) around longitude -180/180 (not 360). So, to avoid spurious drawings and coordinate duplication, we'll _always_ have to patch the geojson after the topojson is fetched in plotly.js.

---

Thinking about this more deeply, perhaps we should instead 1) map longitudes of antimeridian crossing polygons using `lon => Math.sin(lon)` (or `cos` :slightly_smiling_face: ) in `choropleth/plot.js` 2) add a flag in their corresponding `calcdata` items 3) apply `Math.sin` to the lon/lat mouse position in `choropleth/hover.js` when computing the distance from these flagged calcdata items. This might be more robust than the ad-hoc offset or stitch techniques of 190c733. Thoughts?

Reply:
For the record (a conversation @etpinard and I had elsewhere) I worry that `sin` or `cos` mapping would introduce other artifacts, as it maps two `lon` values onto the same output.

I think we're always going to need special handling of some sort or other, because the globe simply doesn't map onto flat space as assumed by `polygon.contains`. That could either be as you've introduced here, or making a special `polygon.contains` just for lat/lon, with special handling for crossing the antimeridian (with additional edge cases for encircling either pole).

Given that at present we only have three entities that can trigger this, I think the solution you have here is perfect. If we extend beyond this in the future (one tricky one that occurs to me is if we allow filled areas in `scattergeo` with `hoveron: 'fills'`) we can generalize at that point.